File: codifide/projection/cbor.py
"""Canonical CBOR projection.

RFC 8949 (CBOR) §4.2 deterministically-encoded subset. Two implementations
of this subset — Python here, Rust in ``crates/codifide-canonical/src/cbor.rs``
— MUST produce byte-identical output for every input; the conformance test
enforces that.

Rules we implement (RFC 8949 §4.2):

- Integers use the shortest major-type-0/1 head. Integers outside
  ``[-2^64, 2^64)`` use tagged bignum (tag 2 for positive, tag 3 for
  negative) with a shortest-byte-string body.
- Floats use the shortest IEEE-754 binary representation that preserves
  their value: half precision if exact, else single, else double. NaN and
  infinity MUST NOT appear in a canonical document and raise ``ValueError``.
- Byte and text strings use the shortest length prefix.
- Arrays use definite-length encoding.
- Maps use definite-length encoding; keys are sorted by their encoded
  bytes in bytewise lexicographic order (RFC 8949 §4.2.1).
- ``None`` encodes as null (major type 7, additional 22).
- ``True`` / ``False`` encode as major type 7, additional 21 / 20.

Nothing else encodes. Unknown Python types raise ``TypeError``. The
canonical form is the contract; ambiguity is a bug.
"""
from __future__ import annotations

import hashlib
import math
import struct
from typing import Any, Optional


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------


def canonical_cbor(value: Any) -> bytes:
    """Encode a JSON-compatible value as deterministic CBOR.

    Accepts ``None``, ``bool``, ``int``, ``float``, ``str``, ``bytes``,
    ``list``, and ``dict`` (with string keys, matching what the canonical
    JSON projection produces). Anything else raises ``TypeError``.
    """
    # ``bool`` is a subclass of ``int`` in Python, so it comes first.
    if value is None:
        return _FIXED_NULL
    if value is True:
        return _FIXED_TRUE
    if value is False:
        return _FIXED_FALSE
    if isinstance(value, bool):
        # Defensive: caught above, but keep the branch explicit.
        return _FIXED_TRUE if value else _FIXED_FALSE
    if isinstance(value, int):
        return _encode_int(value)
    if isinstance(value, float):
        return _encode_float(value)
    if isinstance(value, str):
        return _encode_text(value)
    if isinstance(value, (bytes, bytearray)):
        return _encode_bytes(bytes(value))
    if isinstance(value, list):
        return _encode_array(value)
    if isinstance(value, tuple):
        # Tuples serialize as arrays; no separate CBOR type exists.
        return _encode_array(list(value))
    if isinstance(value, dict):
        return _encode_map(value)
    raise TypeError(f"cannot CBOR-encode {type(value).__name__}: {value!r}")


def content_hash_cbor_bytes(data: bytes) -> str:
    """SHA-256 over canonical CBOR bytes, formatted as ``sha256:<hex>``."""
    return f"sha256:{hashlib.sha256(data).hexdigest()}"


# ---------------------------------------------------------------------------
# Constants
# ---------------------------------------------------------------------------


_FIXED_FALSE = bytes([0xF4])
_FIXED_TRUE = bytes([0xF5])
_FIXED_NULL = bytes([0xF6])


# ---------------------------------------------------------------------------
# Head encoding
# ---------------------------------------------------------------------------


def _encode_head(major: int, n: int) -> bytes:
    """Encode an (major-type, argument) head using the shortest form.

    ``major`` is the 3-bit major type (0..7). ``n`` is the argument value
    (length for strings/arrays/maps, value for integers). We always emit
    the shortest encoding as RFC 8949 §4.2.1 requires.
    """
    assert 0 <= major <= 7
    prefix = major << 5
    if n < 0:
        raise ValueError(f"head argument must be non-negative, got {n}")
    if n < 24:
        return bytes([prefix | n])
    if n < 0x100:
        return bytes([prefix | 24, n])
    if n < 0x10000:
        return bytes([prefix | 25]) + n.to_bytes(2, "big")
    if n < 0x100000000:
        return bytes([prefix | 26]) + n.to_bytes(4, "big")
    if n < 0x10000000000000000:
        return bytes([prefix | 27]) + n.to_bytes(8, "big")
    raise OverflowError(f"CBOR head argument too large: {n}")


# ---------------------------------------------------------------------------
# Integers (including bignums)
# ---------------------------------------------------------------------------


def _encode_int(n: int) -> bytes:
    """Encode an integer, using bignum tags only when out of uint64 range.

    RFC 8949 §4.2.3: integers that fit in a major-type 0/1 head MUST use
    one rather than a bignum tag. We enforce that here.
    """
    if n >= 0:
        if n < 0x10000000000000000:
            return _encode_head(0, n)
        # Positive bignum, tag 2.
        body = _int_to_bytes(n)
        return bytes([0xC2]) + _encode_bytes(body)
    # n < 0: major type 1 encodes -1-n.
    m = -1 - n
    if m < 0x10000000000000000:
        return _encode_head(1, m)
    # Negative bignum, tag 3.
    body = _int_to_bytes(m)
    return bytes([0xC3]) + _encode_bytes(body)


def _int_to_bytes(n: int) -> bytes:
    """Big-endian, shortest unsigned byte representation.

    Zero encodes as a zero-length byte string, which matches what a
    shortest-form encoder would produce (no leading zero bytes). Callers
    of bignum encoding never pass zero — integers of zero fit in a
    single-byte head — but defining the function to handle it keeps the
    contract consistent.
    """
    if n == 0:
        return b""
    return n.to_bytes((n.bit_length() + 7) // 8, "big")


# ---------------------------------------------------------------------------
# Strings and byte strings
# ---------------------------------------------------------------------------


def _encode_text(s: str) -> bytes:
    data = s.encode("utf-8")
    return _encode_head(3, len(data)) + data


def _encode_bytes(b: bytes) -> bytes:
    return _encode_head(2, len(b)) + b


# ---------------------------------------------------------------------------
# Arrays and maps
# ---------------------------------------------------------------------------


def _encode_array(items: list) -> bytes:
    out = bytearray(_encode_head(4, len(items)))
    for item in items:
        out += canonical_cbor(item)
    return bytes(out)


def _encode_map(m: dict) -> bytes:
    """Definite-length map; keys sorted by their encoded bytes.

    RFC 8949 §4.2.1: "The keys in every map MUST be sorted in the
    bytewise lexicographic order of their deterministic encodings."
    Sorting by encoded key rather than by key value matters when keys
    are of mixed types (e.g. integer key 0 vs string key "0"). In the
    canonical Codifide form all keys are strings and the two orderings
    coincide for ASCII keys, but we sort by encoding anyway so the
    implementation is right in general.
    """
    pairs = []
    for k, v in m.items():
        encoded_key = canonical_cbor(k)
        encoded_value = canonical_cbor(v)
        pairs.append((encoded_key, encoded_value))
    pairs.sort(key=lambda p: p[0])
    out = bytearray(_encode_head(5, len(pairs)))
    for ek, ev in pairs:
        out += ek
        out += ev
    return bytes(out)


# ---------------------------------------------------------------------------
# Floats
# ---------------------------------------------------------------------------


def _encode_float(f: float) -> bytes:
    """Encode a float using the shortest IEEE-754 type that preserves it.

    RFC 8949 §4.2.2: "Floating-point values that can be represented as a
    half-precision float MUST be encoded as a half-precision float.
    Floating-point values that cannot be represented as a half-precision
    float but can be represented as a single-precision float MUST be
    encoded as a single-precision float. All other floating-point values
    MUST be encoded as a double-precision float."

    NaN and infinity are not legal in canonical CBOR and raise
    ``ValueError``.
    """
    if math.isnan(f):
        raise ValueError("canonical CBOR: NaN is not representable")
    if math.isinf(f):
        raise ValueError("canonical CBOR: infinity is not representable")

    f16 = _try_half(f)
    if f16 is not None:
        return bytes([0xF9]) + f16

    f32 = _try_single(f)
    if f32 is not None:
        return bytes([0xFA]) + f32

    return bytes([0xFB]) + struct.pack(">d", f)


def _try_half(f: float) -> Optional[bytes]:
    """Return 2-byte IEEE-754 binary16 if it round-trips exactly, else None.

    Python's ``struct`` format ``'e'`` is IEEE-754 binary16, which is
    what CBOR half precision requires. We verify round-trip equality
    because ``struct.pack('>e', ...)`` will silently round values that
    do not fit exactly, and we need exact preservation (or nothing).
    """
    try:
        packed = struct.pack(">e", f)
    except (OverflowError, struct.error):
        return None
    unpacked = struct.unpack(">e", packed)[0]
    # Catch signed zero separately — 0.0 == -0.0 in Python comparison but
    # the sign bit matters for canonical encoding.
    if f == 0.0 and math.copysign(1.0, f) != math.copysign(1.0, unpacked):
        return None
    if unpacked != f:
        return None
    return packed


def _try_single(f: float) -> Optional[bytes]:
    """Return 4-byte IEEE-754 binary32 if it round-trips exactly, else None."""
    try:
        packed = struct.pack(">f", f)
    except (OverflowError, struct.error):
        return None
    unpacked = struct.unpack(">f", packed)[0]
    if f == 0.0 and math.copysign(1.0, f) != math.copysign(1.0, unpacked):
        return None
    if unpacked != f:
        return None
    return packed

File: crates/codifide-canonical/src/cbor.rs
//! Canonical CBOR (RFC 8949 §4.2) byte form.
//!
//! Byte-for-byte compatible with the Python reference implementation in
//! `codifide/projection/cbor.py`. The conformance test pins the two
//! together on every `.cod` example and on a set of RFC 8949 Appendix A
//! test vectors.
//!
//! The encoder accepts the same JSON-compatible subset as the Python
//! side: null, booleans, integers, finite floats, text strings, byte
//! strings, arrays, and maps with string keys. Anything else returns an
//! error. Infinity and NaN are not representable in canonical form.

use std::io::Write;

use serde_json::{Number, Value};
use sha2::{Digest, Sha256};

use crate::error::Error;

/// Produce the canonical CBOR byte form of a JSON-compatible value.
pub fn canonical_cbor(value: &Value) -> Result<Vec<u8>, Error> {
    let mut out = Vec::with_capacity(64);
    encode_value(value, &mut out)?;
    Ok(out)
}

/// SHA-256 of the canonical CBOR byte form, hex-prefixed with `sha256:`.
pub fn content_hash_cbor(value: &Value) -> Result<String, Error> {
    let bytes = canonical_cbor(value)?;
    let mut hasher = Sha256::new();
    hasher.update(&bytes);
    let digest = hasher.finalize();
    let mut hex = String::with_capacity(7 + digest.len() * 2);
    hex.push_str("sha256:");
    for b in digest {
        hex.push(HEX[(b >> 4) as usize] as char);
        hex.push(HEX[(b & 0x0f) as usize] as char);
    }
    Ok(hex)
}

const HEX: &[u8] = b"0123456789abcdef";

// ---------------------------------------------------------------------------
// Core dispatch
// ---------------------------------------------------------------------------

fn encode_value(v: &Value, out: &mut Vec<u8>) -> Result<(), Error> {
    match v {
        Value::Null => {
            out.push(0xF6);
            Ok(())
        }
        Value::Bool(true) => {
            out.push(0xF5);
            Ok(())
        }
        Value::Bool(false) => {
            out.push(0xF4);
            Ok(())
        }
        Value::Number(n) => encode_number(n, out),
        Value::String(s) => encode_text(s, out),
        Value::Array(a) => encode_array(a, out),
        Value::Object(o) => encode_map(o, out),
    }
}

// ---------------------------------------------------------------------------
// Head encoding
// ---------------------------------------------------------------------------

/// Write the shortest CBOR head for `(major, argument)` to `out`.
fn encode_head(major: u8, n: u64, out: &mut Vec<u8>) {
    debug_assert!(major <= 7);
    let prefix = major << 5;
    if n < 24 {
        out.push(prefix | (n as u8));
    } else if n < 0x100 {
        out.push(prefix | 24);
        out.push(n as u8);
    } else if n < 0x10000 {
        out.push(prefix | 25);
        out.extend_from_slice(&(n as u16).to_be_bytes());
    } else if n < 0x1_0000_0000 {
        out.push(prefix | 26);
        out.extend_from_slice(&(n as u32).to_be_bytes());
    } else {
        out.push(prefix | 27);
        out.extend_from_slice(&n.to_be_bytes());
    }
}

// ---------------------------------------------------------------------------
// Numbers
// ---------------------------------------------------------------------------

fn encode_number(n: &Number, out: &mut Vec<u8>) -> Result<(), Error> {
    // Prefer integer representations when they fit; fall back to float
    // only when the value is genuinely a float. `serde_json::Number`
    // distinguishes these for us.
    if let Some(u) = n.as_u64() {
        encode_unsigned(u, out);
        return Ok(());
    }
    if let Some(i) = n.as_i64() {
        // Covers the negative uint64 range via i64.
        encode_signed(i as i128, out);
        return Ok(());
    }
    // Fall-through: only a float could produce Number that is neither
    // u64 nor i64. `as_f64` always succeeds in that case.
    if let Some(f) = n.as_f64() {
        encode_float(f, out)?;
        return Ok(());
    }
    Err(Error::Shape(format!("unrepresentable number: {n}")))
}

fn encode_unsigned(n: u64, out: &mut Vec<u8>) {
    encode_head(0, n, out);
}

fn encode_signed(n: i128, out: &mut Vec<u8>) {
    if n >= 0 {
        // Fits in u64 because we only reach here via i64.
        encode_head(0, n as u64, out);
    } else {
        // CBOR major type 1 stores `-1 - n` as an unsigned argument.
        let m = (-1 - n) as u64;
        encode_head(1, m, out);
    }
}

fn encode_float(f: f64, out: &mut Vec<u8>) -> Result<(), Error> {
    if f.is_nan() {
        return Err(Error::Shape("canonical CBOR: NaN not representable".into()));
    }
    if f.is_infinite() {
        return Err(Error::Shape(
            "canonical CBOR: infinity not representable".into(),
        ));
    }
    // Try half-precision (binary16) first: the canonical form requires
    // the shortest IEEE encoding that preserves the value exactly,
    // including for signed zero.
    if let Some(h) = try_half(f) {
        out.push(0xF9);
        out.extend_from_slice(&h);
        return Ok(());
    }
    if let Some(s) = try_single(f) {
        out.push(0xFA);
        out.extend_from_slice(&s);
        return Ok(());
    }
    out.push(0xFB);
    out.extend_from_slice(&f.to_be_bytes());
    Ok(())
}

/// Return the 2-byte binary16 representation iff it is exact (preserves
/// value and sign of zero), else None.
fn try_half(f: f64) -> Option<[u8; 2]> {
    // Convert via f32 round-trip first: many values that fit in f16
    // also fit in f32, and the f32 path is the definitive check for
    // "narrower float suffices". We then narrow from f32 to f16 by
    // bit-twiddling because Rust has no stable f16 in 1.95.
    let packed = f64_to_f16_bits(f)?;
    Some(packed.to_be_bytes())
}

/// Try to encode `f` as IEEE-754 binary16 bits, returning None if the
/// encoding would lose precision or change the sign of zero.
fn f64_to_f16_bits(f: f64) -> Option<u16> {
    // Handle zero (preserves sign via the magnitude path below).
    if f == 0.0 {
        let sign = (f.to_bits() >> 63) as u16;
        return Some(sign << 15);
    }

    let bits = f.to_bits();
    let sign = ((bits >> 63) & 0x1) as u16;
    let exp_f64 = ((bits >> 52) & 0x7FF) as i32; // 0..2047
    let mant_f64 = bits & 0x000F_FFFF_FFFF_FFFF;

    // NaN/inf were already rejected by the caller.
    let unbiased_exp = exp_f64 - 1023;

    // Normalized representable range: unbiased exponent in [-14, 15]
    // after normalization. Use conservative bounds before fine checks.
    if unbiased_exp > 15 {
        // Overflow for f16 normals (>= 2^16 magnitude).
        return None;
    }

    if unbiased_exp >= -14 {
        // Normal number in f16. The 10-bit mantissa must be obtained by
        // shifting the 52-bit f64 mantissa right by 42 bits and losing
        // no set bits in the low 42 bits (otherwise we'd lose
        // precision).
        if mant_f64 & ((1u64 << 42) - 1) != 0 {
            return None;
        }
        let mant_f16 = (mant_f64 >> 42) as u16;
        let exp_f16 = ((unbiased_exp + 15) as u16) & 0x1F;
        return Some((sign << 15) | (exp_f16 << 10) | mant_f16);
    }

    // Subnormal range for f16: unbiased_exp in [-24, -15]. The f16
    // subnormal representation stores mantissa with an implicit
    // exponent of -14, so we have to re-express f's magnitude in that
    // frame and confirm exactness.
    if unbiased_exp < -24 {
        // Too small even for a subnormal.
        return None;
    }

    // Effective mantissa including the implicit leading 1.
    let mant_with_leading: u64 = mant_f64 | (1u64 << 52);
    // Shift to place at 10-bit mantissa with implicit -14 exponent.
    // The shift amount is: (52 - 10) + (-14 - unbiased_exp) = 42 + (-14 - unbiased_exp).
    let shift = 42 + (-14 - unbiased_exp) as u32;
    // A shift of 64 or more discards all bits; values that require
    // such a shift cannot be represented.
    if shift >= 64 {
        return None;
    }
    // The low `shift` bits must be zero for exactness.
    if mant_with_leading & ((1u64 << shift) - 1) != 0 {
        return None;
    }
    let mant_f16 = (mant_with_leading >> shift) as u16;
    // For subnormals the f16 exponent field is zero.
    Some((sign << 15) | mant_f16)
}

/// Return the 4-byte binary32 representation iff it is exact.
fn try_single(f: f64) -> Option<[u8; 4]> {
    let as_f32 = f as f32;
    if (as_f32 as f64) != f {
        return None;
    }
    // Preserve sign of zero.
    if f == 0.0 && f.is_sign_negative() != as_f32.is_sign_negative() {
        return None;
    }
    Some(as_f32.to_be_bytes())
}

// ---------------------------------------------------------------------------
// Strings, arrays, maps
// ---------------------------------------------------------------------------

fn encode_text(s: &str, out: &mut Vec<u8>) -> Result<(), Error> {
    let bytes = s.as_bytes();
    encode_head(3, bytes.len() as u64, out);
    out.write_all(bytes).unwrap();
    Ok(())
}

fn encode_array(items: &[Value], out: &mut Vec<u8>) -> Result<(), Error> {
    encode_head(4, items.len() as u64, out);
    for item in items {
        encode_value(item, out)?;
    }
    Ok(())
}

fn encode_map(obj: &serde_json::Map<String, Value>, out: &mut Vec<u8>) -> Result<(), Error> {
    // Encode each (key, value) pair, sort by encoded key bytes, emit.
    // Sorting by encoded bytes is what RFC 8949 §4.2.1 requires; it
    // happens to coincide with sorting by unescaped UTF-8 string bytes
    // for string keys, but we do it the general way anyway.
    let mut pairs: Vec<(Vec<u8>, Vec<u8>)> = Vec::with_capacity(obj.len());
    for (k, v) in obj {
        let mut key_buf = Vec::new();
        encode_text(k, &mut key_buf)?;
        let mut val_buf = Vec::new();
        encode_value(v, &mut val_buf)?;
        pairs.push((key_buf, val_buf));
    }
    pairs.sort_by(|a, b| a.0.cmp(&b.0));
    encode_head(5, pairs.len() as u64, out);
    for (k, v) in pairs {
        out.extend_from_slice(&k);
        out.extend_from_slice(&v);
    }
    Ok(())
}

// ---------------------------------------------------------------------------
// Tests (RFC 8949 Appendix A vectors)
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::{json, Number};

    /// Helper: encode a JSON value and return its hex for readable compares.
    fn hex(v: &Value) -> String {
        let b = canonical_cbor(v).unwrap();
        b.iter().map(|x| format!("{x:02x}")).collect()
    }

    #[test]
    fn rfc_8949_integer_vectors() {
        assert_eq!(hex(&json!(0)), "00");
        assert_eq!(hex(&json!(1)), "01");
        assert_eq!(hex(&json!(10)), "0a");
        assert_eq!(hex(&json!(23)), "17");
        assert_eq!(hex(&json!(24)), "1818");
        assert_eq!(hex(&json!(100)), "1864");
        assert_eq!(hex(&json!(1000)), "1903e8");
        assert_eq!(hex(&json!(1_000_000)), "1a000f4240");
        assert_eq!(hex(&json!(1_000_000_000_000u64)), "1b000000e8d4a51000");
        assert_eq!(hex(&json!(-1)), "20");
        assert_eq!(hex(&json!(-10)), "29");
        assert_eq!(hex(&json!(-100)), "3863");
        assert_eq!(hex(&json!(-1000)), "3903e7");
    }

    #[test]
    fn rfc_8949_float_vectors() {
        // 0.0 and -0.0: half-precision preserves signed zero.
        assert_eq!(hex(&Value::Number(Number::from_f64(0.0).unwrap())), "f90000");
        assert_eq!(
            hex(&Value::Number(Number::from_f64(-0.0).unwrap())),
            "f98000"
        );
        assert_eq!(hex(&Value::Number(Number::from_f64(1.0).unwrap())), "f93c00");
        assert_eq!(hex(&Value::Number(Number::from_f64(1.5).unwrap())), "f93e00");
        // 1.1 is not exact in any binary float; encoded as f64.
        assert_eq!(
            hex(&Value::Number(Number::from_f64(1.1).unwrap())),
            "fb3ff199999999999a"
        );
        // 65504.0 fits exactly in half precision (largest subnormal-less normal).
        assert_eq!(
            hex(&Value::Number(Number::from_f64(65504.0).unwrap())),
            "f97bff"
        );
        // 100000.0 does not fit in half but fits in single.
        assert_eq!(
            hex(&Value::Number(Number::from_f64(100000.0).unwrap())),
            "fa47c35000"
        );
        // Smallest f16 subnormal.
        assert_eq!(
            hex(&Value::Number(Number::from_f64(5.960464477539063e-8).unwrap())),
            "f90001"
        );
    }

    #[test]
    fn rfc_8949_simple_and_composite_vectors() {
        assert_eq!(hex(&json!(false)), "f4");
        assert_eq!(hex(&json!(true)), "f5");
        assert_eq!(hex(&json!(null)), "f6");
        assert_eq!(hex(&json!([])), "80");
        assert_eq!(hex(&json!([1, 2, 3])), "83010203");
        assert_eq!(hex(&json!({})), "a0");
        assert_eq!(hex(&json!("")), "60");
        assert_eq!(hex(&json!("a")), "6161");
        assert_eq!(hex(&json!("IETF")), "6449455446");
        assert_eq!(hex(&json!("\u{00fc}")), "62c3bc");
        assert_eq!(hex(&json!("\u{6c34}")), "63e6b0b4");
    }

    #[test]
    fn map_keys_sort_by_encoded_bytes() {
        // Keys of different lengths: CBOR sorts by encoded bytes, so
        // "b" (head 0x61 "text string len 1") sorts before "aa" (head
        // 0x62 "text string len 2") because 0x61 < 0x62 regardless of
        // the body. The canonical encoder must emit them in that order.
        let v = json!({"aa": 2, "b": 1});
        let got = hex(&v);
        // 0xa2 (map(2)), then key "b" = 61 62, value 01,
        // then key "aa" = 62 61 61, value 02.
        assert_eq!(got, "a26162016261610 2".replace(" ", ""));
    }
}
